realm/realm-library/src/main/cpp/io_realm_RealmFunctions.cpp

```diff
@@ -18,19 +18,40 @@
 
 #include "util.hpp"
 #include "jni_util/bson_util.hpp"
+#include "java_network_transport.hpp"
+#include "object-store/src/sync/app.hpp"
 
 using namespace realm;
+using namespace realm::app;
 using namespace realm::jni_util;
 
-// FIXME This is just a basic round trip test for passing bson back and forth. Proper implementation
-//  will come with actual Function implementation.
-JNIEXPORT jstring JNICALL Java_io_realm_RealmFunctions_nativeCallFunction
-        (JNIEnv* env, jclass, jstring j_args) {
+static std::function<jobject(JNIEnv*, Optional<bson::Bson> )> response_mapper = [](JNIEnv* env, Optional<bson::Bson> response) {
+    if (response) {
+        return JniBsonProtocol::bson_to_jstring(env, *response);
+    } else {
+        // FIXME How to raise errors here
```

Comment:
If you throw a `std::logic_error` or `std:runtime_error` they will be caught by the `CATCH()` macro in the JNI layer and converted into a Java exception. 

The logic for how the conversion is defined here: https://github.com/realm/realm-java/blob/v10/realm/realm-library/src/main/cpp/util.cpp#L156

If those exceptions do not match the error you want to return to Java, we normally create a custom exception class in C++ that can be caught explicitly and converted.

Reply:
Seem like we should never reach here, as we would not reach here if we do not have a valid Bson response. I added a `std::logic_error` to propagate it nicely should it ever happen.